social_django/views.py

```diff
@@ -39,21 +43,82 @@ def disconnect(request, backend, association_id=None):
                          redirect_name=REDIRECT_FIELD_NAME)
 
 
+def get_session_timeout(
+        social_user,
+        enable_session_expiration=False,
+        max_session_length=None,
+):
+    if enable_session_expiration:
+        # Retrieve an expiration date from the social user who just finished
+        # logging in; this value was set by the social auth backend, and was
+        # typically received from the server.
+        expiration = social_user.expiration_datetime()
+
+        # We've enabled session expiration. Check to see if we got
+        # a specific expiration time from the provider for this user;
+        # if not, use the platform default expiration.
+        received_expiration_time = (
+            expiration.seconds + (expiration.days * 86400)
+        ) if expiration else DEFAULT_SESSION_TIMEOUT
+
+        # Check to see if the backend set a value as a maximum length
+        # that a session may be; if they did, then we should use the minimum
+        # of that and the received session expiration time, if any, to
+        # set the session length.
+        if received_expiration_time is None and max_session_length is None:
+            # We neither received an expiration length, nor have a maximum
+            # session length. Use the platform default.
+            session_expiry = DEFAULT_SESSION_TIMEOUT
+        elif max_session_length is not None and received_expiration_time is None:
+            # We only have a maximum session length; use that.
+            session_expiry = max_session_length
+        elif received_expiration_time is not None and max_session_length is None:
+            # We only have an expiration time received by the backend
+            # from the provider, with no set maximum. Use that.
+            session_expiry = received_expiration_time
+        else:
+            # We received an expiration time from the backend, and we also
+            # have a set maximum session length. Use the smaller of the two.
+            session_expiry = min(received_expiration_time, max_session_length)
```

Comment:
@haikuginger Thanks for documenting this control flow extensively :)

One suggestion regarding readability: It would be least surprising if the code were checking values of `received_expiration_time` and `max_session_length` in the same order every time. That is:

```python
        if received_expiration_time is None and max_session_length is None:
            # ...
        elif received_expiration_time is None and max_session_length is not None:
            # ...
        elif received_expiration_time is not None and max_session_length is None:
            # ...
        else:
            # ...
```